File: try_java/src/main/java/com/example/model/DefaultUserRepository.java
package com.example.model;

import io.micronaut.transaction.annotation.ReadOnly;
import jakarta.inject.Singleton;

import javax.persistence.EntityManager;
import javax.persistence.TypedQuery;
import javax.transaction.Transactional;
import javax.validation.constraints.NotBlank;
import java.util.List;
import java.util.Optional;

@Singleton
public class DefaultUserRepository implements UserRepository {
    private final EntityManager entityManager;

    public DefaultUserRepository(EntityManager entityManager) {
        this.entityManager = entityManager;
    }

    @Override
    @ReadOnly//All database access needs to be wrapped inside a transaction. As the method only reads data from the database, annotate it with @ReadOnly.
    public Optional<UserDto> findById(long id) {
        return Optional.ofNullable(entityManager.find(UserDto.class, id));
    }

    @Override
    @Transactional//This method modifies the database, thus it is annotated with @Transactional.
    public UserDto save(@NotBlank String name) {
        UserDto userDto = new UserDto();
        userDto.setUsername(name);
        userDto.setPassword("pwd");
        entityManager.persist(userDto);
        return userDto;
    }

    @Override
    @ReadOnly//All database access needs to be wrapped inside a transaction. As the method only reads data from the database, annotate it with @ReadOnly.
    public List<UserDto> findAll() {
        String qlString = "SELECT u FROM UserDto as u";
        TypedQuery<UserDto> query = entityManager.createQuery(qlString, UserDto.class);
        return query.getResultList();
    }
}

File: try_kotlin/src/main/kotlin/com/example/model/DefaultUserRepository.kt
package com.example.model

import io.micronaut.transaction.annotation.ReadOnly
import jakarta.inject.Singleton
import javax.persistence.EntityManager
import javax.persistence.PersistenceContext
import javax.transaction.Transactional

@Singleton
open class DefaultUserRepository(@PersistenceContext private val entityManager: EntityManager): UserRepository {

    @ReadOnly //All database access needs to be wrapped inside a transaction. As the method only reads data from the database, annotate it with @ReadOnly.
    override fun findById(id: Long): UserDto? {
        return entityManager.find(UserDto::class.java, id)
    }

    @Transactional //This method modifies the database, thus it is annotated with @Transactional.
    override fun save(name: String): UserDto {
        val userDto = UserDto()
        userDto.username = name
        userDto.password = "pwd"
        entityManager.persist(userDto)
        return userDto
    }

    @ReadOnly //All database access needs to be wrapped inside a transaction. As the method only reads data from the database, annotate it with @ReadOnly.
    override fun findAll(): List<UserDto>? {
        val qlString = "SELECT u FROM UserDto as u"
        val query = entityManager.createQuery(qlString, UserDto::class.java)
        return query.resultList
    }
}